rebalance.py

```diff
@@ -335,25 +334,88 @@ def _name_total_gen_htmiss_prior_pdf(self):
     def _name_total_gen_htmiss_prior_slope(self):
         return 'gen_htmiss_prior_slope'
 
-    def _build_gen_htmiss_prior(self):
-        slope_name = self._name_total_gen_htmiss_prior_slope()
-        slope_variable = r.RooRealVar(
-            slope_name,
-            slope_name,
-            -0.05,
-        )
-        self._wsimp(slope_variable)
+    def _get_gen_htmiss_prior_file(self):
+        rfile = './input/htmiss_prior.root'
+        return r.TFile(rfile)
 
-        prior_pdf_name = self._name_total_gen_htmiss_prior_pdf()
+    def _name_ht_bin(self, gen_ht):
+        '''Given the GEN-HT of the event, figure out which HT bin it corresponds to.'''
+        # Binning of the prior in terms of HT
+        htbins = [100, 300, 500, 700, 900, 1300, 2000, 5000]
+        for idx in range(len(htbins)-1):
+            lo = htbins[idx]
+            hi = htbins[idx+1]
+
+            if (lo <= gen_ht) and (gen_ht <= hi):
+                return f'{lo:.0f}_to_{hi:.0f}'
+
+        raise RuntimeError(f'Could not figure out the HT bin for HT: {gen_ht:.3f}')
+
+    def _get_prior_histogram(self, key):
+        hist = key.ReadObj()
+        histname = hist.GetName()
+        return hist, histname
+
+    def _convert_th1_to_roohistpdf_and_save_to_ws(self, th1, th1name):
+        '''Do the TH1 -> RooDataHist -> RooHistPdf conversion. Saves the final RooHistPdf into the workspace.'''
         htmiss_variable = self.ws.function(self._name_partial_gen_htmiss_variable(direction='pt'))
-        prior_pdf = r.RooExponential(
-            prior_pdf_name,
-            prior_pdf_name,
-            htmiss_variable,
-            slope_variable
+
+        # We need to provide at least one RooRealVar (primitive variable) into the RooDataHist and RooHistPdf
+        # Providing only derived quantities do not work in the datahist/pdf constructors
+        # To achieve that, we define this dummy RooRealVar and evaluate it at the initial value of HTmiss.
+        dummy_htmiss_variable = r.RooRealVar(
+            'dummy_gen_htmiss_pt',
+            'dummy_gen_htmiss_pt',
+            htmiss_variable.evaluate()
         )
+
+        datahist = r.RooDataHist(th1name, th1name,
+                        r.RooArgList(dummy_htmiss_variable),
+                        th1
+                    )
+
+        prior_pdf_name = self._name_total_gen_htmiss_prior_pdf()
+
+        prior_pdf = r.RooHistPdf(prior_pdf_name,
+                        prior_pdf_name,
+                        r.RooArgList(htmiss_variable),
+                        r.RooArgList(dummy_htmiss_variable),
+                        datahist
+                    )
+
+        # Linear interpolation
+        prior_pdf.setInterpolationOrder(1)
+
         self._wsimp(prior_pdf)
```

Comment:
Just as a quick comment about a detail here:

I re-factored the code that does TH1 -> RooHistPdf and saves the pdf into the workspace all in one function. If I try to return the pdf to the main function, and attempt to save it there, I ran into seg fault issues, so I moved the saving part into the same function as well.

Reply:
Thanks for clarifying. That's fine.
Two things though:
1. the name of the function sounds very general; one might think that one can use it for all kinds of different roohistpdf creation. However, the naming and htmiss variable choice as input to the roohistpdf are pre-set and cannot be changed, so this function can really only be used for one thing: Creating the ht prior pdf. Suggest to rename to '_build_gen_htmiss_prior_roohistpdf'.
2. What is the point of the `th1name` argument? Seems like a) it doesn't matter in the end, bc the roodatahist is thrown away anyway and b) we currently just put in `th1name=th1.GetName()`, so I do not understand why it needs to be a separate argument

Reply:
Thanks for the suggestions, function name is updated and th1name argument is removed in a1f7933